unit_test/test_tag_trimming.py

```diff
@@ -0,0 +1,55 @@
+import unittest, nttt
+
+class test_tag_trimming(unittest.TestCase):
```

Comment:
The following test case fails:

```
    def test_spaced_tag_content_legitimate(self):
        init = "In XML every opening tag (<foo>) should have a matching closing tag (</foo>)"

        out = "In XML every opening tag (<foo>) should have a matching closing tag (</foo>)"

        self.assertEqual(nttt.tag_trimming.trim_tags(init, 'any_lang'), out)
```

In general this kind of cleanup is quite dangerous since it may lead to some nasty false positives. And getting those false positives is even worse than not having the check at all, because it may spoil perfect translations.

So I'd suggest to re-evaluate this requirement with Nina and Maja highlighting all possible risks.

Reply:
Great idea! If I'm given a comprehensive list of examples that mustn't be changed it shouldn't be difficult to work with those restrictions given the implementation of the script.

Forgetting the false positives I would love to know whether you think the general style to be good as I would be thrilled to tweak it may it be needed.

I have also seen your message on the Slack channel. I believe waiting for Nina's and Maja's input is the most sensible thing to do.

Feel free to close this PR. I'll otherwise close tomorrow it as this code is not to be integrated whatsoever.

Thanks a lot for the review and sorry for all the hassle...

Reply:
I'm no Python expert, in my day job I'm a Java software engineer, so I can only hope that my suggestions are not complete nonsense :)

I really like that you have quite comprehensive set of unit tests. But it's really hard to grasp what's going on here:

```

    if '/' in match:
        if match[-2] != ' ' and match[-1] != '>' and ((match[-1] != '-' and t_lang in hypen_langs) or t_lang not in hypen_langs):
            match = match[:-1] + ' ' + match[-1]
        if match[1] == ' ':
            match = match[0] + match[1:].lstrip()
    else:
        if match[1] == '<':
            match = match[0] + ' ' + match[1:]
        if match[-2] == ' ':
            match = match[:-1].rstrip() + match[-1
```

Some sensible method names could probably help.

I'd like to know about the comprehensive list you mentioned but parsing natural human language with regular expressions is unfortunately prone to unexpected errors. So at this point I think that it would be better to have some semi-automatic script that asks humans for confirmations for these kinds of corrections. On the other hand I perfectly understand that it changes the implementation approach  to a great extent. 

If you have any ideas on how to make these kinds of corrections safe and automatic please shout.

Reply:
Thanks a lot for the comments! I'll break the code further down from now on. I usually tend to try and cram as much as possible on as few lines as I can and that has proven to be counterproductive more often than not...

I have tweaked the code so as to pass the unit test you proposed. I'll make a new commit so that you can take a look at how I've solved it.

If we are to apply automatic correction I bet the most sensible option would be to put together a broad list of cases to look out for and then perform a lot of tests again said scenarios. This would in turn require close cooperation with Nina and Maja. Nonetheless, I'm more keen on the semi-automatic approach. As soon as we settle on something I'll fix this issue up. I'll close this PR for now though.

Thanks again for the comments!